File: langs/zig/src/parser.zig
//! Clide usage string parser

const std = @import("std");
const spec = @import("spec.zig");

pub const ParseError = error{
    NoUsageLines,
    InvalidUsageLine,
    UnterminatedPositional,
    BadOption,
    BadPositional,
    UnknownType,
    OutOfMemory,
};

pub fn fromLines(allocator: std.mem.Allocator, lines: []const []const u8) ParseError!spec.Spec {
    if (lines.len == 0) {
        return error.NoUsageLines;
    }

    var parsed = std.ArrayList(struct { []const u8, []spec.Item }).init(allocator);
    defer parsed.deinit();

    for (lines) |line| {
        const parsed_line = try parseUsageLine(allocator, line);
        try parsed.append(parsed_line);
    }

    const prog = parsed.items[0].@"0";

    var commands = std.ArrayList(spec.Command).init(allocator);
    for (parsed.items) |item| {
        const name = firstLitAfterProg(item.@"1") orelse "_";
        const name_owned = try allocator.dupe(u8, name);
        try commands.append(.{
            .name = name_owned,
            .items = item.@"1",
        });
    }

    const prog_owned = try allocator.dupe(u8, prog);
    const commands_owned = try commands.toOwnedSlice();

    return spec.Spec{
        .prog = prog_owned,
        .commands = commands_owned,
    };
}

fn parseUsageLine(allocator: std.mem.Allocator, line: []const u8) ParseError!struct { []const u8, []spec.Item } {
    var tokens = std.ArrayList([]const u8).init(allocator);
    defer tokens.deinit();

    var iter = std.mem.tokenizeScalar(u8, line, ' ');
    while (iter.next()) |token| {
        try tokens.append(token);
    }

    if (tokens.items.len < 2 or !std.mem.eql(u8, tokens.items[0], "Usage:")) {
        return error.InvalidUsageLine;
    }

    const prog = tokens.items[1];
    var items = std.ArrayList(spec.Item).init(allocator);

    for (tokens.items[2..]) |token| {
        const item = try parseGroupToken(allocator, token);
        try items.append(item);
    }

    return .{ prog, try items.toOwnedSlice() };
}

fn parseGroupToken(allocator: std.mem.Allocator, token: []const u8) ParseError!spec.Item {
    if (token.len >= 2 and token[0] == '[' and token[token.len - 1] == ']') {
        const inner = token[1..(token.len - 1)];
        var alts = std.ArrayList(spec.Atom).init(allocator);

        var iter = std.mem.splitScalar(u8, inner, '|');
        while (iter.next()) |alt| {
            const atom = try atomOf(allocator, alt);
            try alts.append(atom);
        }

        const atoms = try alts.toOwnedSlice();
        const group = if (atoms.len == 1) blk: {
            const single_atom = atoms[0];
            allocator.free(atoms);
            break :blk spec.Group{ .single = single_atom };
        } else
            spec.Group{ .alt = atoms };

        return spec.Item{ .optional = group };
    } else {
        const atom = try atomOf(allocator, token);
        const group = spec.Group{ .single = atom };
        return spec.Item{ .required = group };
    }
}

fn atomOf(allocator: std.mem.Allocator, token: []const u8) ParseError!spec.Atom {
    _ = allocator;
    if (token.len > 0 and token[0] == '<') {
        if (token.len < 2 or token[token.len - 1] != '>') {
            return error.UnterminatedPositional;
        }
        return parsePos(token);
    } else if (isLong(token) or isShort(token) or std.mem.indexOfScalar(u8, token, '=') != null) {
        return parseOption(token);
    } else {
        return spec.Atom{ .lit = token };
    }
}

fn parseOption(token: []const u8) ParseError!spec.Atom {
    const eq_pos = std.mem.indexOfScalar(u8, token, '=');
    const name_part = if (eq_pos) |pos| token[0..pos] else token;
    const val_part = if (eq_pos) |pos| token[(pos + 1)..] else null;

    const long_opt: ?[]const u8 = if (isLong(name_part)) name_part else null;
    const short_opt: ?[]const u8 = if (isShort(name_part)) name_part else null;

    if (long_opt == null and short_opt == null) {
        return error.BadOption;
    }

    if (val_part == null) {
        return spec.Atom{
            .opt_bool = .{
                .long = long_opt,
                .short = short_opt,
            },
        };
    }

    const allow_repeat = val_part.?.len > 0 and val_part.?[val_part.?.len - 1] == '+';
    const core = if (allow_repeat) val_part.?[0..(val_part.?.len - 1)] else val_part.?;

    const colon_pos = std.mem.indexOfScalar(u8, core, ':');
    const ty_str = if (colon_pos) |pos| core[0..pos] else core;
    const default = if (colon_pos) |pos| core[(pos + 1)..] else null;

    const ty = try spec.Type.fromStr(ty_str);

    return spec.Atom{
        .opt_val = .{
            .long = long_opt,
            .short = short_opt,
            .ty = ty,
            .default = default,
            .allow_repeat = allow_repeat,
        },
    };
}

fn parsePos(token: []const u8) ParseError!spec.Atom {
    if (token.len < 2 or token[0] != '<' or token[token.len - 1] != '>') {
        return error.BadPositional;
    }

    const inside = token[1..(token.len - 1)];
    const colon1 = std.mem.indexOfScalar(u8, inside, ':') orelse return error.BadPositional;
    const colon2 = std.mem.indexOfScalarPos(u8, inside, colon1 + 1, ':');

    const name = inside[0..colon1];
    const ty_str = if (colon2) |c2| inside[(colon1 + 1)..c2] else inside[(colon1 + 1)..];
    const default = if (colon2) |c2| inside[(c2 + 1)..] else null;

    const ty = try spec.Type.fromStr(ty_str);
    return spec.Atom{
        .pos = .{
            .name = name,
            .ty = ty,
            .default = default,
        },
    };
}

fn isShort(s: []const u8) bool {
    return s.len == 2 and s[0] == '-' and s[1] != '-';
}

fn isLong(s: []const u8) bool {
    return s.len >= 3 and s[0] == '-' and s[1] == '-';
}

fn firstLitAfterProg(items: []const spec.Item) ?[]const u8 {
    for (items) |item| {
        switch (item) {
            .required => |group| {
                switch (group) {
                    .single => |atom| {
                        switch (atom) {
                            .lit => |lit| return lit,
                            else => {},
                        }
                    },
                    .alt => |atoms| {
                        for (atoms) |atom| {
                            switch (atom) {
                                .lit => |lit| return lit,
                                else => {},
                            }
                        }
                    },
                }
            },
            .optional => {},
        }
    }
    return null;
}

File: langs/rust/src/parser.rs
//! Clide usage string parser

use crate::spec::*;

#[derive(Debug)]
pub struct ParseError(pub String);

pub fn from_lines(lines: &[String]) -> Result<Spec, ParseError> {
    if lines.is_empty() {
        return Err(ParseError("No usage lines".to_string()));
    }

    let parsed: Result<Vec<_>, _> = lines.iter().map(|s| parse_usage_line(s)).collect();
    let parsed = parsed?;

    let prog = parsed[0].0.clone();

    let commands: Vec<Command> = parsed
        .into_iter()
        .map(|(_, items)| {
            let name = first_lit_after_prog(&items).unwrap_or("_".to_string());
            Command { name, items }
        })
        .collect();

    Ok(Spec { prog, commands })
}

fn parse_usage_line(line: &str) -> Result<(String, Vec<Item>), ParseError> {
    let words: Vec<&str> = line.split_whitespace().collect();
    
    if words.len() < 2 || words[0] != "Usage:" {
        return Err(ParseError("Line must start with 'Usage:'".to_string()));
    }

    let prog = words[1].to_string();
    let rest: Vec<&str> = words[2..].to_vec();
    let items: Result<Vec<Item>, _> = rest.iter().map(|t| parse_group_token(t)).collect();
    let items = items?;

    Ok((prog, items))
}

fn parse_group_token(token: &str) -> Result<Item, ParseError> {
    if token.starts_with('[') && token.ends_with(']') {
        let inner = &token[1..token.len() - 1];
        let alts: Vec<&str> = inner.split('|').collect();
        
        let atoms: Result<Vec<Atom>, _> = alts.iter().map(|a| atom_of(a)).collect();
        let atoms = atoms?;

        let group = if atoms.len() == 1 {
            Group::Single(atoms.into_iter().next().unwrap())
        } else {
            Group::Alt(atoms)
        };

        Ok(Item::Optional(group))
    } else {
        let atom = atom_of(token)?;
        Ok(Item::Required(Group::Single(atom)))
    }
}

fn atom_of(token: &str) -> Result<Atom, ParseError> {
    if token.starts_with('<') {
        if !token.ends_with('>') {
            return Err(ParseError(format!("Unterminated positional: {}", token)));
        }
        parse_pos(token)
    } else if is_long(token) || is_short(token) || token.contains('=') {
        parse_option(token)
    } else {
        Ok(Atom::Lit(token.to_string()))
    }
}

fn parse_option(token: &str) -> Result<Atom, ParseError> {
    let (name_part, val_part) = if let Some(pos) = token.find('=') {
        let (n, v) = token.split_at(pos);
        (n, Some(&v[1..]))
    } else {
        (token, None)
    };

    let (long_opt, short_opt) = if is_long(name_part) {
        (Some(name_part.to_string()), None)
    } else if is_short(name_part) {
        (None, Some(name_part.to_string()))
    } else {
        return Err(ParseError(format!("Bad option: {}", token)));
    };

    match val_part {
        None => Ok(Atom::OptBool {
            long: long_opt,
            short: short_opt,
        }),
        Some(v) => {
            let allow_repeat = v.ends_with('+');
            let core = if allow_repeat {
                &v[..v.len() - 1]
            } else {
                v
            };

            let (ty_str, default) = if let Some(pos) = core.find(':') {
                let (t, d) = core.split_at(pos);
                (t, Some(&d[1..]))
            } else {
                (core, None)
            };

            let ty = Type::from_str(ty_str)
                .map_err(|e| ParseError(format!("{}: {}", e, token)))?;

            Ok(Atom::OptVal {
                long: long_opt,
                short: short_opt,
                ty,
                default: default.map(|s| s.to_string()),
                allow_repeat,
            })
        }
    }
}

fn parse_pos(token: &str) -> Result<Atom, ParseError> {
    if !token.starts_with('<') || !token.ends_with('>') {
        return Err(ParseError(format!("Expected <...>: {}", token)));
    }

    let inside = &token[1..token.len() - 1];
    let parts: Vec<&str> = inside.split(':').collect();

    match parts.as_slice() {
        [nm, tys] => {
            let ty = Type::from_str(tys)
                .map_err(|e| ParseError(format!("{}: {}", e, token)))?;
            Ok(Atom::Pos {
                name: nm.to_string(),
                ty,
                default: None,
            })
        }
        [nm, tys, def] => {
            let ty = Type::from_str(tys)
                .map_err(|e| ParseError(format!("{}: {}", e, token)))?;
            Ok(Atom::Pos {
                name: nm.to_string(),
                ty,
                default: Some(def.to_string()),
            })
        }
        _ => Err(ParseError(format!("Bad positional: {}", token))),
    }
}

fn is_short(s: &str) -> bool {
    s.len() == 2 && s.starts_with('-') && !s.starts_with("--")
}

fn is_long(s: &str) -> bool {
    s.len() >= 3 && s.starts_with("--")
}

fn first_lit_after_prog(items: &[Item]) -> Option<String> {
    for item in items {
        if let Item::Required(group) = item {
            match group {
                Group::Single(Atom::Lit(s)) => return Some(s.clone()),
                Group::Alt(atoms) => {
                    if let Some(Atom::Lit(s)) = atoms.first() {
                        return Some(s.clone());
                    }
                }
                _ => {}
            }
        }
    }
    None
}
